prov/usnic/src/usdf_cq.c

```diff
@@ -119,7 +119,55 @@ usdf_cq_readerr(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
 
 	cq->cq_comp.uc_status = 0;
 
-	return sizeof(*entry);
+	return 1;
+}
+
+static ssize_t
+usdf_cq_readerr_soft(struct fid_cq *fcq, struct fi_cq_err_entry *entry,
+		uint64_t flags)
+{
+	struct usdf_cq *cq;
+	size_t soft_cq_entry_len;
+	struct usdf_cq_soft_entry *soft_comp_entry;
+	void *tail;
+
+	USDF_TRACE_SYS(CQ, "\n");
+
+	cq = container_of(fcq, struct usdf_cq, cq_fid);
+
+	// The return values are analogous to sockets cq_readerr
+	tail = cq->c.soft.cq_tail;
+	soft_cq_entry_len = sizeof(struct usdf_cq_soft_entry);
+	soft_comp_entry = (struct usdf_cq_soft_entry *)tail;
+
+	entry->op_context = soft_comp_entry->op_context;
+	entry->flags = 0;
+	entry->err = FI_EIO;
+	switch (soft_comp_entry->status) {
+	case USD_COMPSTAT_SUCCESS:
+		entry->prov_errno = FI_SUCCESS;
+		break;
+	case USD_COMPSTAT_ERROR_CRC:
+		entry->prov_errno = FI_ECRC;
+		break;
+	case USD_COMPSTAT_ERROR_TRUNC:
+		entry->prov_errno = FI_ETRUNC;
+		break;
+	case USD_COMPSTAT_ERROR_TIMEOUT:
+		entry->prov_errno = FI_ETIMEDOUT;
+		break;
+	case USD_COMPSTAT_ERROR_INTERNAL:
+		entry->prov_errno = FI_EOTHER;
+		break;
+	}
+
+	tail = (uint8_t *)tail + soft_cq_entry_len;
```

Comment:
> If I'm changing cq_tail to struct usdf_cq_soft_entry * , shouldn't I change cq_comps, cq_end, cq_head also?

Yes, that was implied.

> If yes, then I will have to make sure, the change remains consistent across soft cq and hard cq, as these variables are used in usdf_progress_hard_cq.

Yes you will.  Though I don't think that `usdf_progress_hard_cq` is actually ever called right now, since both the EP_MSG and EP_RDM code replace the `cqh_progress` function pointer with `usdf_msg_hcq_progress` and `usdf_rdm_hcq_progress` respectively.